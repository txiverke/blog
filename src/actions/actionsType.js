// @flow

export const LOAD_STATS_REQUEST = 'LOAD_STATS_REQUEST'
export const LOAD_STATS_SUCCESS = 'LOAD_STATS_SUCCESS'
export const LOAD_STATS_FAILURE = 'LOAD_STATS_FAILURE'

export const LOG_IN_USER_REQUEST = 'LOG_IN_USER_REQUEST'
export const LOG_IN_USER_SUCCESS = 'LOG_IN_USER_SUCCESS'
export const LOG_IN_USER_FAILURE = 'LOG_IN_USER_FAILURE'

export const IS_AUTHENTICATED_REQUEST = 'IS_AUTHENTICATE_REQUEST'
export const IS_AUTHENTICATED_SUCCESS = 'IS_AUTHENTICATE_SUCCESS'
export const IS_AUTHENTICATED_FAILURE = 'IS_AUTHENTICATE_FAILURE'

export const LOAD_USER_DATA_REQUEST = 'LOAD_USER_DATA_REQUEST'
export const LOAD_USER_DATA_SUCCESS = 'LOAD_USER_DATA_SUCCESS'
export const LOAD_USER_DATA_FAILURE = 'LOAD_USER_DATA_FAILURE'

export const UPDATE_USER_DATA_REQUEST = 'UPDATE_USER_DATA_REQUEST'
export const UPDATE_USER_DATA_SUCCESS = 'UPDATE_USER_DATA_SUCCESS'
export const UPDATE_USER_DATA_FAILURE = 'UPDATE_USER_DATA_FAILURE'

export const LOG_OUT_USER_REQUEST = 'LOG_OUT_USER_REQUEST'
export const LOG_OUT_USER_SUCCESS = 'LOG_OUT_USER_SUCCESS'
export const LOG_OUT_USER_FAILURE = 'LOG_OUT_USER_FAILURE'

export const LOAD_POST_DATA_REQUEST = 'LOAD_POST_DATA_REQUEST'
export const LOAD_POST_DATA_SUCCESS = 'LOAD_POST_DATA_SUCCESS'
export const LOAD_POST_DATA_FAILURE = 'LOAD_POST_DATA_FAILURE'

export const LOAD_POST_ITEM_REQUEST = 'LOAD_POST_ITEM_REQUEST'
export const LOAD_POST_ITEM_SUCCESS = 'LOAD_POST_ITEM_SUCCESS'
export const LOAD_POST_ITEM_FAILURE = 'LOAD_POST_ITEM_FAILURE'
export const UPDATE_POST_ITEM_SUCCESS = 'UPDATE_POST_ITEM_SUCCESS'

export const CREATE_POST_DATA_REQUEST = 'CREATE_POST_DATA_REQUEST'
export const CREATE_POST_DATA_SUCCESS = 'CREATE_POST_DATA_SUCCESS'
export const CREATE_POST_DATA_FAILURE = 'CREATE_POST_DATA_FAILURE'

export const UPDATE_POST_DATA_REQUEST = 'UPDATE_POST_DATA_REQUEST'
export const UPDATE_POST_DATA_SUCCESS = 'UPDATE_POST_DATA_SUCCESS'
export const UPDATE_POST_DATA_FAILURE = 'UPDATE_POST_DATA_FAILURE'

export const REMOVE_POST_DATA_REQUEST = 'REMOVE_POST_DATA_REQUEST'
export const REMOVE_POST_DATA_SUCCESS = 'REMOVE_POST_DATA_SUCCESS'
export const REMOVE_POST_DATA_FAILURE = 'REMOVE_POST_DATA_FAILURE'
