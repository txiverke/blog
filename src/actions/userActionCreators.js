// flow

import * as ACTION from './actionsType'
import config from '../config'
import { handleToken } from '../utils/helpers'

/** LOGIN USER **/
export const loginUserRequest = () => ({ type: ACTION.LOG_IN_USER_REQUEST })
export const loginUserSuccess = (payload: Object) => ({ type: ACTION.LOG_IN_USER_SUCCESS, payload })
export const loginUserFailure = () => ({ type: ACTION.LOG_IN_USER_FAILURE})

export const loginUser = (obj: Object) =>
  async (dispatch: Function) => {
    const headers = new Headers({ 'Content-Type': 'application/json' })

    dispatch(loginUserRequest())

    try {
      const promise = await fetch(config.api.auth, { 
        method: 'POST', 
        body: JSON.stringify(obj), 
        headers 
      })
      const data = await promise.json()
      handleToken.tokens = JSON.stringify(data.token)
      dispatch(loginUserSuccess(data.token))
      dispatch(isAuthenticatedSuccess(data.token))
      return dispatch(loadUserData(config.api.profileId))
    } catch (err) {
      console.log(err)
      dispatch(loginUserFailure())
    }
  }

/** AUTHENTICATE USER  */
export const isAuthenticatedRequest = () => ({ type: ACTION.IS_AUTHENTICATED_REQUEST })
export const isAuthenticatedSuccess = (payload: string) => ({ type: ACTION.IS_AUTHENTICATED_SUCCESS, payload })
export const isAuthenticatedFailure = () => ({ type: ACTION.IS_AUTHENTICATED_FAILURE})

export const isAuthenticated = () => 
  (dispatch: Function) => {
    handleToken.tokens
    const token: ?string = handleToken.tokens
    dispatch(isAuthenticatedRequest())

    if (token) {
      dispatch(isAuthenticatedSuccess(token))
      return dispatch(loadUserData(config.api.profileId))
    }

    return dispatch(isAuthenticatedFailure())
  }

/** LOGOUT USER */
export const logoutUserRequest = () => ({ type: ACTION.LOG_OUT_USER_REQUEST })
export const logoutUserSuccess = () => ({ type: ACTION.LOG_OUT_USER_SUCCESS })
export const logoutUserFailure = () => ({ type: ACTION.LOG_OUT_USER_FAILURE })

export const logoutUser = () => (dispatch: Function) => {
  const token = handleToken.tokens
  dispatch(logoutUserRequest())

  if (token) {
    handleToken.remove()
    dispatch(logoutUserSuccess())
    return dispatch((isAuthenticated()))
  }

  return dispatch(logoutUserFailure())
}

/** LOAD USER DATA **/
export const loadUserDataRequest = () => ({ type: ACTION.LOAD_USER_DATA_REQUEST })
export const loadUserDataSuccess = (payload: Object) => ({ type: ACTION.LOAD_USER_DATA_SUCCESS, payload })
export const loadUserDataFailure = () => ({ type: ACTION.LOAD_USER_DATA_FAILURE })

export const loadUserData = (id: string) =>
  async (dispatch: Function) => {
    const token: ?string = handleToken.tokens
    const headers = new Headers({
      'Content-Type': 'application/json',
      'access_token': token || ''
    })
    const options = { 
      method: 'GET',
      headers,
      mode: 'cors',
      cache: 'default',
    }

    dispatch(loadUserDataRequest())

    try {
      const promise = await fetch(`${config.api.url}/users/${id}`, options)
      const data = await promise.json()
      return dispatch(loadUserDataSuccess(data))
    } catch (err) {
      return dispatch(loadUserDataFailure())
    }
  }

/** UPLOAD USER DATA **/
export const uploadUserDataRequest = () => ({ type: ACTION.UPLOAD_USER_DATA_REQUEST })
export const uploadUserDataSuccess = (payload: Object) => ({ type: ACTION.UPLOAD_USER_DATA_SUCCESS, payload })
export const uploadUserDataFailure = () => ({ type: ACTION.UPLOAD_USER_DATA_FAILURE })

export const uploadUserData = (id: string, obj: Object) =>
  async (dispatch: Function) => {
    const token: ?string = handleToken.tokens
    const headers = new Headers({
      'Content-Type': 'application/json',
      'access_token': token || ''
    })
    const options = { 
      method: 'PUT',
      body: JSON.stringify(obj),
      headers,
      mode: 'cors',
      cache: 'default',
    }

    dispatch(uploadUserDataRequest())

    try {
      const promise = await fetch(`${config.api.url}/users/${id}`, options)
      const data = await promise.json()
      return dispatch(uploadUserDataSuccess(data))
    } catch (err) {
      return dispatch(uploadUserDataFailure())
    }
  }