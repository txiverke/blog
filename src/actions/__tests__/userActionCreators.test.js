// @flow

import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import * as action from '../userActionCreators'
import * as type from '../actionsType'

const middlewares = [ thunk ]
const mockStore = configureMockStore(middlewares)

describe('userActionCreators', () => {
  describe('login user', () => {
    it('should return LOGIN_USER_REQUEST action', () => {
      expect(action.loginUserRequest()).toMatchSnapshot()
    })

    it('should return LOGIN_USER_SUCCESS action', () => {
      expect(action.loginUserSuccess({})).toMatchSnapshot()
    })

    it('should return LOGIN_USER_FAILURE action', () => {
      expect(action.loginUserFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})
      const expectedActions = [
        'LOG_IN_USER_REQUEST',
        'LOG_IN_USER_FAILURE'
      ]

      return store.dispatch(action.loginUser({}))
        .then(() => {
          const actualActions = store.getActions().map(action => action.type)
          expect(actualActions).toEqual(expectedActions)
       })
    })
  })

  describe('authenticate user', () => {
    it('should return IS_AUTHENTICATE_REQUEST action', () => {
      expect(action.isAuthenticatedRequest()).toMatchSnapshot()
    })

    it('should return IS_AUTHENTICATE_SUCCESS action', () => {
      expect(action.isAuthenticatedSuccess('')).toMatchSnapshot()
    })

    it('should return IS_AUTHENTICATE_FAILURE action', () => {
      expect(action.isAuthenticatedFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})
      const expectedActions = [
        'IS_AUTHENTICATE_REQUEST',
        'IS_AUTHENTICATE_FAILURE'
      ]

      store.dispatch(action.isAuthenticated())
      const actualActions = store.getActions().map(action => action.type)
      expect(actualActions).toEqual(expectedActions)
    })
  })

  describe('logout user', () => {
    it('should return LOG_OUT_USER_REQUEST action', () => {
      expect(action.logoutUserRequest()).toMatchSnapshot()
    })

    it('should return LOG_OUT_USER_SUCCESS action', () => {
      expect(action.logoutUserSuccess('')).toMatchSnapshot()
    })

    it('should return LOG_OUT_USER_FAILURE action', () => {
      expect(action.logoutUserFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})
      const expectedActions = [
        'LOG_OUT_USER_REQUEST',
        'LOG_OUT_USER_FAILURE'
      ]

      store.dispatch(action.logoutUser())
      const actualActions = store.getActions().map(action => action.type)
      expect(actualActions).toEqual(expectedActions)
    })
  })

  describe('load user data', () => {
    it('should return LOAD_USER_DATA_REQUEST action', () => {
      expect(action.loadUserDataRequest()).toMatchSnapshot()
    })

    it('should return LOAD_USER_DATA_SUCCESS action', () => {
      expect(action.loadUserDataSuccess({})).toMatchSnapshot()
    })

    it('should return LOAD_USER_DATA_FAILURE action', () => {
      expect(action.loadUserDataFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})
      const expectedActions = [
        'LOAD_USER_DATA_REQUEST',
        'LOAD_USER_DATA_FAILURE'
      ]

      return store.dispatch(action.loadUserData(''))
      .then(() => {
        const actualActions = store.getActions().map(action => action.type)
        expect(actualActions).toEqual(expectedActions)
     })
    })
  })

  describe('update user data', () => {
    it('should return UPDATE_USER_DATA_REQUEST action', () => {
      expect(action.updateUserDataRequest()).toMatchSnapshot()
    })

    it('should return UPDATE_USER_DATA_SUCCESS action', () => {
      expect(action.updateUserDataSuccess({})).toMatchSnapshot()
    })

    it('should return UPDATE_USER_DATA_FAILURE action', () => {
      expect(action.updateUserDataFailure()).toMatchSnapshot()
    })

    it('expected actions should be dispatched on request', () => {
      const store = mockStore({})
      const expectedActions = [
        'UPDATE_USER_DATA_REQUEST',
        'UPDATE_USER_DATA_FAILURE'
      ]

      return store.dispatch(action.updateUserData({}))
      .then(() => {
        const actualActions = store.getActions().map(action => action.type)
        expect(actualActions).toEqual(expectedActions)
     })
    })
  })
})