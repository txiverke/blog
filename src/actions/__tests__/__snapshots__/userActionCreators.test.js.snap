// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userActionCreators authenticate user should return IS_AUTHENTICATE_FAILURE action 1`] = `
Object {
  "type": "IS_AUTHENTICATE_FAILURE",
}
`;

exports[`userActionCreators authenticate user should return IS_AUTHENTICATE_REQUEST action 1`] = `
Object {
  "type": "IS_AUTHENTICATE_REQUEST",
}
`;

exports[`userActionCreators authenticate user should return IS_AUTHENTICATE_SUCCESS action 1`] = `
Object {
  "payload": "",
  "type": "IS_AUTHENTICATE_SUCCESS",
}
`;

exports[`userActionCreators load user data should return LOAD_USER_DATA_FAILURE action 1`] = `
Object {
  "type": "LOAD_USER_DATA_FAILURE",
}
`;

exports[`userActionCreators load user data should return LOAD_USER_DATA_REQUEST action 1`] = `
Object {
  "type": "LOAD_USER_DATA_REQUEST",
}
`;

exports[`userActionCreators load user data should return LOAD_USER_DATA_SUCCESS action 1`] = `
Object {
  "payload": Object {},
  "type": "LOAD_USER_DATA_SUCCESS",
}
`;

exports[`userActionCreators login user should return LOGIN_USER_FAILURE action 1`] = `
Object {
  "type": "LOG_IN_USER_FAILURE",
}
`;

exports[`userActionCreators login user should return LOGIN_USER_REQUEST action 1`] = `
Object {
  "type": "LOG_IN_USER_REQUEST",
}
`;

exports[`userActionCreators login user should return LOGIN_USER_SUCCESS action 1`] = `
Object {
  "payload": Object {},
  "type": "LOG_IN_USER_SUCCESS",
}
`;

exports[`userActionCreators logout user should return LOG_OUT_USER_FAILURE action 1`] = `
Object {
  "type": "LOG_OUT_USER_FAILURE",
}
`;

exports[`userActionCreators logout user should return LOG_OUT_USER_REQUEST action 1`] = `
Object {
  "type": "LOG_OUT_USER_REQUEST",
}
`;

exports[`userActionCreators logout user should return LOG_OUT_USER_SUCCESS action 1`] = `
Object {
  "type": "LOG_OUT_USER_SUCCESS",
}
`;

exports[`userActionCreators update user data should return UPDATE_USER_DATA_FAILURE action 1`] = `
Object {
  "type": "UPDATE_USER_DATA_FAILURE",
}
`;

exports[`userActionCreators update user data should return UPDATE_USER_DATA_REQUEST action 1`] = `
Object {
  "type": "UPDATE_USER_DATA_REQUEST",
}
`;

exports[`userActionCreators update user data should return UPDATE_USER_DATA_SUCCESS action 1`] = `
Object {
  "payload": Object {},
  "type": "UPDATE_USER_DATA_SUCCESS",
}
`;
